### PR TITLE
[quant][graph] Add quantized batch_norm2d_relu to graph mode

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36622 [quant][graph] Graph mode quantization support for sigmoid
* **#36552 [quant][graph] Add quantized batch_norm2d_relu to graph mode**
* #36692 [quant][graph] Add quantized batch_norm2d support to graph mode
* #36691 [quant][graph] Add useQuantizable function

Summary:
Do the fusion for inplace and non-inplace relu
Tested for functional relu as well.
Functional batch_norm is not a usual use-case (since it expects the weight, bias, mean, var) so that is not tested.

Test Plan:
test_quantize_script.py test_batch_norm2d_relu

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21075253](https://our.internmc.facebook.com/intern/diff/D21075253)

### CODE REVIEW

File: test/quantization/test_quantize_script.py [hunk 1]
```diff
@@ -1405,6 +1405,42 @@ def forward(self, x):
...
+            def forward(self, x):
+                return self.relu(self.bn(x))
+
+        class Func(torch.nn.Module):
+            def __init__(self):
+                super(Func, self).__init__()
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
+
+            def forward(self, x):
+                return F.relu(self.bn(x))
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1424]: can you add a `inplace` argument here as well?

File: test/quantization/test_quantize_script.py [hunk 2]
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
...
+            def forward(self, x):
+                return self.relu(self.bn(x))
+
+        class FuncRelu(torch.nn.Module):
+            def __init__(self):
+                super(FuncRelu, self).__init__()
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
+
+            def forward(self, x):
+                return F.relu(self.bn(x), True)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1424]: can you make inplace an argument to the module?

File: test/quantization/test_quantize_script.py [hunk 3]
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
...
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
+                self.relu = torch.nn.ReLU(inplace=inplace)
+
+            def forward(self, x):
+                return self.relu(self.bn(x))
+
+        class FuncRelu(torch.nn.Module):
+            def __init__(self):
+                super(FuncRelu, self).__init__()
+                self.bn = torch.nn.BatchNorm2d(3).to('cpu', torch.float)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1421]: `cpu` is default

File: test/quantization/test_quantize_script.py [hunk 4]
```diff
@@ -1405,6 +1405,50 @@ def forward(self, x):
...
+            m = prepare_script(m, {'': default_qconfig}, True)
+            data = torch.rand((1, 3, 10, 10), dtype=torch.float)
+            m(data)
+            m = convert_script(m, False)
+            FileCheck().check("quantized::batch_norm2d_relu") \
+                       .check_not("aten::batch_norm") \
+                       .check_not("aten::relu") \
+                       .check_not("aten::relu_") \
+                       .run(m.graph)
+
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1451]: can you remove code duplication here:?

File: test/quantization/test_quantize_script.py (hunk 1 above)
Comment (supriyar) [reply test/quantization/test_quantize_script.py:1425]: Added a comment explaining that it is not supported.

Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1425]: you can still test for the case the `inplace` argument is a constant I think

File: test/quantization/test_quantize_script.py [hunk 5]
```diff
@@ -1405,6 +1405,42 @@ def forward(self, x):
...
+        model = self._test_op_impl(FuncRelu, data, "quantized::batch_norm2d_relu")
+        FileCheck().check_not("aten::batch_norm") \
+                   .check_not("aten::relu") \
+                   .run(model.graph)
+        for inplace in [True, False]:
+            m = torch.jit.script(M(inplace).eval())
+            m = prepare_script(m, {'': default_qconfig}, True)
+            data = torch.rand((1, 3, 10, 10), dtype=torch.float)
+            m(data)
+            m = convert_script(m, False)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1437]: use `_test_op_impl`?

Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1437]: can you merge all the checks in one for loop?

File: test/quantization/test_quantize_script.py [hunk 6]
```diff
@@ -1163,9 +1163,9 @@ class TestQuantizeScriptPTSQOps(JitTestCase):
     """ Test graph mode post training static quantization works
     for individual ops end to end.
     """
-    def _test_op_impl(self, SingleOpModule, data, quantized_op):
+    def _test_op_impl(self, SingleOpModule, data, quantized_op, **kwargs):
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1166]: can you change this to take a module instance? then you can initialize the module before passing into this function.

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
please refactor `_test_op_impl`

File: test/quantization/test_quantize_script.py [hunk 7]
```diff
@@ -1400,11 +1400,47 @@ def forward(self, x):
...
+                super(BNFuncInplaceRelu, self).__init__()
+                self.bn = torch.nn.BatchNorm2d(3).to(torch.float)
+
+            def forward(self, x):
+                return F.relu(self.bn(x), True)
+
+        data = [(torch.rand((1, 3, 10, 10), dtype=torch.float), torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        for Model in [BNRelu, BNFuncRelu, BNFuncInplaceRelu]:
+            for inplace in [True, False]:
+                model = self._test_op_impl(Model(inplace), data, "quantized::batch_norm2d_relu")
```
Comment (jerryzh168) [inline test/quantization/test_quantize_script.py:1438]: please change this loop as well, we don't need to loop through [True, False] anymore

Comment (supriyar) [reply test/quantization/test_quantize_script.py:1438]: We still need to test BNRelu model for different inplace argument.

Comment (supriyar) [reply test/quantization/test_quantize_script.py:1438]: Good catch! Updated it

**Review by jerryzh168 — APPROVED** [review]
Looks good, thanks!
